rfcs/20221006-frontend_event_tracking.md

```diff
@@ -0,0 +1,469 @@
+
+## The technical details
+
+We will create a system for collecting events on the frontend. Each event will have a unique name, collect some default data about the enviroment (timestamp, referer, and so on) and can have an optional data payload containing string and numerical values. I'd like to limit the payload shape to strings and numbers to keep it JSON serializable and simple. Allowing for complex nested objects might prove too complex to query in the future. I think this is a reasonable constraint but am open to alternatives.
+
+### Sending events
+
+These events will be sent as `POST` requests to a server route on the frontend, which will then route them to our desired analytics service. In the future, it may make sense to move this server route to a standalone microservice, but for now a [Nuxt server route](https://v3.nuxtjs.org/guide/directory-structure/server/#server-routes) or [Server Middleware](https://nuxtjs.org/docs/configuration-glossary/configuration-servermiddleware/) (once we're updated to Nuxt 3), should be sufficient to handle the analytics events. This endpoint is essentially a proxy between the frontend and our 'true' analytics service that makes sure we're always sending events server-side.
+
+#### Privacy
+
+Events will _only_ be collected and sent under the following conditions:
+
+- An `ANALYTICS_ENABLED` feature flag is set to `true`.
+- The user has not opted out of analytics by setting the `allowAnalytics` value in our UI cookie store to `false`.
+
+Unless all conditions are met the `sendEvent` function will simply no-op.
+
+A previous version of this proposal suggested honoring the `navigator.doNotTrack` browser method, but upon inspection this was deprecated in 2018 and is no longer advised.
+
+This proposal will require the creation of some new user interface elements to toggle analytics tracking:
+
+1. A small popup banner that shows up to users, which informs them that Openverse collects anonymized analytics information to make the service better. There are "dismiss" and "learn more" calls to action. "Learn more" directs users to:
+2. An analytics opt-out checkbox on the `/privacy` page. We also add information describing our analytics collection, and link to the relevant code in our GitHub repository.
+
+### Events
+
+Below I've created a TypeScript type for all the proposed anlytics events, and a simple function for sending events. For each event you'll see the payload alongside the default values which will be passed to _all_ events.
+
+```ts
+// src/utils/events.ts
+
+// This type is only for illustrative purposes and would be imported from elsewhere.
+type MediaType = "image" | "audio" | "all";
```

Comment:
Since it include "all, it should be `SearchType` (`MediaType` is for the actual types of media, not the aggregate "All content")